docs: add docs for `@sanity/client/csm` and `@sanity/client/stega` (#377)
sanity-io/client: README.md

Content Source Maps annotate fragments in your query results with metadata about its origin: the field, document, and dataset it originated from.

> [!IMPORTANT]  

> > Content Source Maps are supported only in versions `2021-03-25` or later of the Content Lake API.

Reviewer: ```suggestion
>
> Content Source Maps are supported in the Content Lake API versions `2021-03-25` and later.
```

> Read the [Content Source Maps introduction][content-source-maps-intro] before diving in, and keep the [Content Source Maps reference][content-source-maps] handy.

Reviewer: ```suggestion
Before diving in, review the [Content Source Maps introduction][content-source-maps-intro] and keep the [Content Source Maps reference][content-source-maps] within reach for a quick lookup.
```

   console.log(resultSourceMap)
   ```

> Once enabled, the `resultSourceMap` property will always exist on the response, given your `apiVersion` is `2021-03-25` or later. If there is no source map, it will be an empty object. There's also a TypeScript definition for it:

Reviewer: ```suggestion
If your `apiVersion` is `2021-03-25` or later,  the `resultSourceMap` property will always exist in the response after enabling it. If there is no source map, `resultSourceMap` is an empty object.  
This is the corresponding TypeScript definition:
```

useContentSourceMap(resultSourceMap)
```

#### Using [Visual editing][visual-editing] with steganography

A turnkey integration with [Visual editing][visual-editing] is available in [`@sanity/client/stega`]. It creates edit intent links for all the string values in your query result, using [steganography](https://npmjs.com/package/@vercel/stega) under the hood.

```ts
import {createClient} from '@sanity/client/stega'

const client = createClient({
  // ...base config options
  stega: {

>     // If you use Vercel Visual Editing then it makes sense to always enable it on Preview deployments

Reviewer: ```suggestion
    // If you use Vercel Visual Editing, we recommend enabling it for Preview deployments
```

    enabled: process.env.VERCEL_ENV === 'preview',

>     // Required: Set it to relative or absolute URL of your Sanity Studio

Reviewer: ```suggestion
    // Required: Set it to the relative or absolute URL of your Sanity Studio instance
```

    studioUrl: '/studio', // or 'https://your-project-name.sanity.studio'

>     // For resolving Cross Dataset References you can pass a function returning a URL

Reviewer: ```suggestion
    // To resolve Cross Dataset References, pass a function returning a URL
```

    studioUrl: (sourceDocument: ContentSourceMapDocument | ContentSourceMapRemoteDocument) => {

>       // If `sourceDocument` has a projectId and dataset, then it's a Cross Dataset Reference

Reviewer: ```suggestion
      // If `sourceDocument` has a projectId and a dataset, then it's a Cross Dataset Reference
```

      if (source._projectId && source._dataset) {
        return 'https://acme-global.sanity.studio'
      }
      return 'https://acme-store.sanity.studio'
    },
    // If your Studio has Workspaces: https://www.sanity.io/docs/workspaces

>     // And your Cross Dataset References are available in a workspace then you can return an object to let the client setup the URL

Reviewer: ```suggestion
    // and if your Cross Dataset References are available in a workspace, you can return an object to let the client set up the URL
```

    studioUrl: (sourceDocument) => {
      // This organization has a single studio with everything organized in workspaces
      const baseUrl = 'https://acme.sanity.studio'

>       // If `sourceDocument` has a projectId and dataset, then it's a Cross Dataset Reference

Reviewer: ```suggestion
      // If `sourceDocument` has a projectId and a dataset, then it's a Cross Dataset Reference
```

      if (source._projectId && source._dataset) {
        return {baseUrl, workspace: 'global'}
      }
      return {baseUrl, workspace: 'store'}
    },

>     // Optional, control what fields have stega payloads

Reviewer: ```suggestion
    // Optional, to control which fields have stega payloads
```

    filter: (props) => {
      const {resultPath, sourcePath, sourceDocument, value} = props
      if (sourcePath[0] === 'externalurl') {
        return false
      }

>       // The default behavior is packaged into `filterDefault`, allowing you enable encoding fields that are skipped by default

Reviewer: ```suggestion
      // The default behavior is packaged into `filterDefault`, allowing you to enable encoding fields that are skipped by default
```

      return props.filterDefault(props)
    },

>     // Optional, logs what's encoded and what isn't

Reviewer: ```suggestion
    // Optional, to log what's encoded and what isn't
```

```ts
import {createClient} from '@sanity/client' // or '@sanity/client/stega'
import {resolveEditUrl} from '@sanity/client/csm'

const client = createClient({
  // ... standard client config
  // Required: the new 'withKeyArraySelector' option is used here instead of 'true' so that links to array items and portable text are stable even if the array is reordered
  resultSourceMap: 'withKeyArraySelector',
})
const {result, resultSourceMap} = await client.fetch(
  `*[_type == "author" && slug.current == $slug][0]{name, pictures}`,
  {slug: 'john-doe'},
  // Required, otherwise you can't access `resultSourceMap`
  {filterResponse: false},
)

// The `result` looks like this:
const result = {
  name: 'John Doe',
  pictures: [
    {
      _type: 'image',
      alt: 'A picture of exactly what you think someone named John Doe would look like',
      _key: 'cee5fbb69da2',
      asset: {
        _ref: 'image-a75b03fdd5b5fa36947bf2b776a542e0c940f682-1000x1500-jpg',
        _type: 'reference',
      },
    },
  ],
}

const studioUrl = 'https://your-project-name.sanity.studio'

resolveEditUrl({

>   // The URL resolver needs to know the base URL of your Sanity Studio

Reviewer: ```suggestion
  // The URL resolver requires the base URL of your Sanity Studio instance
```

  studioUrl,

>   // And a Content Source Map for the query result you want to create an edit intent link for

Reviewer: ```suggestion
  // It also requires a Content Source Map for the query result you want to create an edit intent link for
```

  resultSourceMap,

>   // The path to the field you want to edit, you can pass a string

Reviewer: ```suggestion
  // The path to the field you want to edit. You can pass a string
```

  resultPath: 'pictures[0].alt',

>   // Or an array of segments

Reviewer: ```suggestion
  // or an array of segments
```
